design/m1_push_to_file_design.md

```diff
@@ -0,0 +1,1247 @@
+Details are TBD, but a representative configuration might look something like
+this:
+
+- 5 secrets groups
+- 5 secrets / secrets group
+- Mix of secret value lengths (10 to 20 characters, plus some base64-encoded
+  TLS certificates)
+
+## Logs
+
+| **Scenario** | **Log Level** | **Notes** |
+| ------------ | ------------- | --------- |
+| Annotation has `conjur.org` prefix, but invalid key or value format | `Error` | Note 1 |
+| Annotation has `conjur.org` prefix, but unknown annotation type | `Info` | Note 2 |
+| Start of secrets retrieval for a group | `Info` | Includes group name |
+| Start of secrets file creation for a group | `Info` | Includes group name |
+| Error processing Go template | `Error` | Note 3 |
+| Error writing secrets file | `Error` | Note 3 |
+
+- *Note 1:* Annotation values are treated as non-sensitive and will be included the logs.
+- *Note 2:* For forward compatability, annotation keys with the prefix
+  `conjur.org` but an unexpected annotation type will be logged at `Info` level.
+- *Note 3:* Care must be taken to prevent secrets file content from leaking.
+
+## Documentation
+
+Documentation for the M1 Secrets Provider Push to File feature will include:
+
+- Update the
+  [Secrets Provider init container configuration documentation](https://docs.cyberark.com/Product-Doc/OnlineHelp/AAM-DAP/Latest/en/Content/Integrations/k8s-ocp/cjr-k8s-secrets-provider-ic.htm).
+- Update the [Secrets Provider application container configuration documentation](https://docs.cyberark.com/Product-Doc/OnlineHelp/AAM-DAP/Latest/en/Content/Integrations/k8s-ocp/cjr-k8s-secrets-provider-ac.htm).
+- Add new documentation to describe the push-to-file annotation configuration
+- Add documentation for the process to upgrade from the Secrets Provider legacy (environment variable based) configuration to the annotation-based configuration.
+
+## Version update
+
+The M1 Push-to-File feature will necessitate a minor version bump for
+the Secrets Provider. Because configuration will maintain backwards
+compatibility, it is not expected that a major version bump will be required.
+
+## Security
+
+### Secrets File Attributes
+
+Since the Push-to-File feature involves writing sensitive information to
+files that are shared by an application container, care needs to be taken
+so that files are created with the proper file attributes:
+
+- File permissions
+- File owner/UID
+- File group/GID
+
+By default, the Secrets Provider will create secrets files with the following
+file attributes, based on the default user defined in the Secrets Provider
+container image:
+
+- File permissions: 660         (`rw-rw----`)
+- File owner: secrets-provider  (UID 777)
+- File group: secrets-provider  (GID 777)
+
+If the application container is built to run with a non-root user that has
+different a UID and GID than the default values shown above, then the
+application will not be able to access secrets files that have been
+created using the default attributes.
```

Comment:
This worries me a bit.  I thought that containers were provided user IDs in OpenShift?  How do we...
1. Reduce the chance that the app cannot read the file.
2. Support the developer in the rare case where their app cannot read the file.
Reading through the sections below... I'm not sure I can quickly grok our options.  Maybe best we chat separately.

Reply:
I reworded this entire section, based on our discussion. Secrets Provider will set file permissions to `664` or `rw-rw-r--` for the secrets files to avoid mismatches when a deployment is configured to run containers in a Pod using an arbitrary UID and GID at run time. The security picture for using these permissions should be analogous to what is done for OpenShift. (I've added all of these details to the Security section.)